dali/pipeline/executor/executor.cc

```diff
@@ -644,7 +649,45 @@ bool Executor<WorkspacePolicy, QueuePolicy>::HasConditionals() const {
   return has_conditionals_;
 }
 
+template<typename WorkspacePolicy, typename QueuePolicy>
+void Executor<WorkspacePolicy, QueuePolicy>::InitCheckpointing() {
+  if (!checkpointing_)
+    return;
+
+  // TODO(mstaniewski): verify
+  int queue_depth = stage_queue_depths_[OpType::CPU] + 1;
+  checkpoint_queue_.resize(queue_depth);
+  for (auto &cpt : checkpoint_queue_)
+    cpt.Build(*graph_);
+
+  checkpointing_epoch_size_ = -1;
+  for (const auto &node : graph_->GetOpNodes()) {
+    auto meta = node.op->GetReaderMeta();
+    if (meta.epoch_size_padded <= 0)
+      continue;
+
+    int local_epoch_size = (meta.epoch_size_padded + max_batch_size_ - 1) / max_batch_size_;
+    if (checkpointing_epoch_size_ == -1)
+      checkpointing_epoch_size_ = local_epoch_size;
+    else if (checkpointing_epoch_size_ != local_epoch_size)
+      DALI_FAIL(make_string(
+        "Incompatibile epoch sizes of operators in a pipeline with checkpointing enabled. ",
+        "Expected: ", checkpointing_epoch_size_, " found: ", local_epoch_size, ". "));
+  }
+}
```

Comment:
1. How about moving the part with epoch to a separate method with a nice TODO comment about it being a limitation/simplification for now, to be lifted later on?
2. What would happen if none of the operators had the ReaderMeta (e.g. a pipeline with just the random number generators)
3. Can you please check how the external_source works here? Does it have the meta? Will it fail on an attempt to get the checkpoint?
4. Nitpick, I think the error message is a bit unclear. How about: "When the checkpointing is enabled, all readers must have the same epoch size. The readers <get the op name> and <get the second op name> have different epoch sizes (<> and <> respecitively)."

Reply:
I am pretty sure that external_source does not have ReaderMeta. I think that it is impossible to make it checkpointable, unless you assume that the input is a pure function, in which case it is stateless, and we have the checkpointing for free.

Am I missing something?

Reply:
Ad 1. Just to have anything that touches the notion of epoch separated and defined in one place. So that we can easily find where they are used and remove the usaged and defnitions when the restriction is ready to be lifted.
Ad 2. It could be either that or an error, I wonder which one is better.
Ad 3. Side note, the external source is special in that regard it allows you to feed smaller batches (than the max_batch_size/batch_size declared in the pipeline_def). What I am asking is not to make it work with checkponting now, but if it fails gracefully. :) I guess if we ever ask it for the checkpoint, the external source op just fails with the unsupported error as any other op. Maybe it's worth to double check though. Especially, as if we go with another behaviour in the 2. point, i.e. we actually never call the SaveCheckpoint, we could end up with reporting to the user that checkpointing is not supported when they try to save the checkpoint (possibly after some heavy data loading and training) rather than just from the start.

Reply:
2. Why would an error be better? A pipeline where the only input is a random generator sound little bit sketchy, but checkpointing it is a valid use case in the end, isn't it?
3. ExternalSource in a pipeline with checkpointint enabled is going to fail right before the first iteration, when the first checkpoint is created, so there shouldn't be any problems. Even though, it seems like a good test case for future python tests.

Reply:
Ad 3. Cool!
Ad 2. You're right, there's no point in making the provided example not work. If the epoch size is -1 we are either able to checkpoint every iteration or none (i.e. external_source). I was worried about the latter case, but we should end up with a proper error in the first iteration.